### DISCUSSION TITLE
Nearby POI selection in navigation

### DISCUSSION THREAD

**Discussion by Ciutnik** [devemux86/cruiser#368] [Q&A] [answered]
Currently, the set of POIs we are informed about in navigation only applies to motor vehicles - cars or motorcycles: fuel/charging station, parking, speed camera. It certainly won't be useful for cyclists and pedestrians (and it may not even be interesting for all motorists).

I understand the current limitation to only four categories (performance), but it would be nice to be able to independently select POIs visible during navigation for each profile independently. Would that be possible?

**Comment by devemux86** [comment]
@Ciutnik indeed we need a function like this for all travel types.

We can define some popular categories and I will include them in the settings.

For example see the POI categories / tags used in the search screen:
- https://github.com/devemux86/cruiser/discussions/195#discussioncomment-6451058

**Reply by Ciutnik** [reply]
> We can define some popular categories and I will include them in the settings.

Do you plan to introduce the ability for the user to change these initial settings according to their preferences? 

For me, it would be ideal to be able to enter your own search key. For example, on a motorcycle, I would like to see a warning before a railway crossing or information about a roadside picnic shelter and viewpoint. However, when walking, in addition to a tourist attraction, I would like to hear about a work of art or a monument... This is all available in POIs, but not in a simple way 🙂

**Reply by devemux86** [reply]
How search with so many possible keys / values could be possible?
We already have predefined categories (as most apps do), what alternative is there?

We can always use key=value in the search field, or only the key or only the value, it will still find results.

**Reply by Ciutnik** [reply]
> We can always use key=value in the search field, or only the key or only the value, it will still find results.

Yes, and it works well! But if I could add this result to the information during navigation, it would be great for me 🙂 Hence my proposal to allow entering the any key value instead of selecting a ready-made POI group. Of course, this is just a loose idea that may not be easy to implement.

In OsmAnd, this is solved in such a way that instead of a dozen or so POI groups, we can see several dozen categories on the map (and in the navigation, ~~as far as I remember~~). It seems to me that we can select any POI included in OSM to be permanently shown on the map as a layer.

**Reply by devemux86** [reply]
This could be possible: search for POI before navigation, use the POI results in navigation's display / announcement.

Then we would not need any POI categories in navigation? Any search result (categories or free search) can be used.

I had also thought about it when I implemented the nearby POI, but preferred to start with a simple implementation
and extend it in the future as we will meet new user requirements, as we are discussing it now. :slightly_smiling_face:

**Reply by Ciutnik** [reply]
Or maybe permanently associate these free searches with profiles? Of course as an option. As I wrote in the first entry, everyone could select four of predefined categories of POI in their profile and/or enter their own key for serach results.

**Reply by devemux86** [reply]
> Or maybe permanently associate these free searches with profiles?

This is also a good idea for nearby POI in navigation, we could store favorite searches or do custom searches.

**Comment by devemux86** [comment]
Next Cruiser will allow to select the nearby POI in navigation between all available POI categories.

The available POI categories are the same as in the search screen. They are mentioned [here](https://github.com/devemux86/cruiser/discussions/195#discussioncomment-6451058).

If we need more detailed POI categories, we can discuss it. e.g. split some of them?
Important: we have to translate the names of the categories in 15 (voice) languages.

<img src="https://github.com/user-attachments/assets/fe41130b-948a-4bea-b3fa-e5c0d779e192" width="50%"/><img src="https://github.com/user-attachments/assets/12aaafd2-4997-4fa5-a7ec-376ba92c0d59" width="50%"/>

**Reply by Ciutnik** [reply]
This news is honey to my heart. And if it were possible to enable these POIs as a dynamic overlay while browsing the map, I could uninstall OsmAnd 😁

**Reply by devemux86** [reply]
@Ciutnik 

> if it were possible to enable these POIs as a dynamic overlay while browsing the map

This will be one of the next big features and will be easier with the new map engine.

Do we have a discussion for this? If not, can you please create a new discussion for it?

**Reply by r11d11** [reply]
POI selection in navigation are very good news

> If we need more detailed POI categories, we can discuss it. e.g. split some of them?

Today Biergarten is included in Drink
I vote for an own entry for Biergarten
In Osmand Biergarten has also its own entry:

![grafik](https://github.com/user-attachments/assets/f56afea1-b1f6-4562-8623-c8b2abf5f82b)

As a motorcyclist I always rather like to sit ouside in a beer garden than in a restaurant
And you can always eat something in a beer garden. Category Drink does only fit to 50%

**Reply by devemux86** [reply]
@r11d11 we can do this.

In navigation settings and search screen.

**Reply by Ciutnik** [reply]
Great. Works as expected. Thanks! 👍

There is only a "clash" between the icon on the map and the icon resulting from the above function. I know that for some reason you want to use Google icons, but maybe it is possible to make the icons resulting from the search as similar as possible to the icons on the map?

For example, a tourist attraction icon: on a map it's a star, and in search results it's a asterisk. But on a map, a asterisk is a viewpoint, etc. 🙂

**Reply by devemux86** [reply]
@Ciutnik 

Important:
Changing map icons is difficult, as it requires changing all the themes that use them.
Also, each theme may use different icons, so we should not expect similarities.

App icons are easier to change, if it is needed.
If there are specific suggestions from [here](https://fonts.google.com/icons?icon.set=Material+Icons&icon.style=Filled), I'll be happy to discuss them. :slightly_smiling_face:

**Reply by Ciutnik** [reply]
Of course I didn't mean changing the icons on the map! 🙂 

I was just thinking about making the search icons and the icons from the theme (Motorider is probably the most popular choice) similar.
Are the theme icons (source: jar) embedded in the app or in the map file?

**Reply by devemux86** [reply]
Motorider and other themes use the same large collection of icons.
You can see it [here](https://github.com/mapsforge/mapsforge/tree/master/mapsforge-themes/src/main/resources/assets/symbols) and is included in the app.

All map icons have been prepared carefully (some time ago) because they need special coloring to be readable.

If you think an icon in Motorider could be changed, but from the existing collection of icons,
please write in the Motorider thread:
- https://github.com/devemux86/cruiser/discussions/612

---

It is often easier to change an app icon, if you can make specific suggestion with the replacement from [here](https://fonts.google.com/icons?icon.set=Material+Icons&icon.style=Filled).

**Comment by devemux86** [comment]
Nearby POI selection in navigation is available for testing in the latest:
- https://github.com/devemux86/cruiser/discussions/610

---

We have separate settings for display panels and voice announcement:
- Settings | Navigation | Panels | Nearby points of interest
- Settings | Voice guidance | Announcement | Nearby points of interest

So we can choose for nearby POI:
- Only display them
- Only announce them
- Display + announce them

As these settings seem like too many, do we need a separate selection for display + announcement?
Or keep the settings only in "Navigation" and announce them anyway when their display is enabled?

**Reply by WolfgangZ42** [reply]
> keep the settings only in "Navigation" and announce them anyway when their display is enabled?

That's okay - a separation isn't necessary - imho.

**Reply by devemux86** [reply]
Even better, we can make the nearby POI voice a single switch, like the nearby favorites.
(it will work for the enabled POI in navigation settings)

**Reply by DolomitesTom** [reply]
I am of the opposite opinion. I want to see and be able to select the nearby POI, but don't want to have them announced. I am for the 3 options mentioned above.

**Reply by devemux86** [reply]
This is only about voice, the detailed POI switches remain in navigation settings.
A POI voice switch turns on / off the POI announcement, like in nearby favorites.

**Reply by DolomitesTom** [reply]
if this is like in b7, it is ok, I thought you want to delete the settings in announcement and announce it as soon as the display is active.

**Reply by devemux86** [reply]
The detailed POI settings in voice guidance were replaced with one switch like the favorites.
The detailed POI selection remains in navigation settings:

<img src="https://github.com/user-attachments/assets/12aaafd2-4997-4fa5-a7ec-376ba92c0d59" width="50%"/><img src="https://github.com/user-attachments/assets/f9a7da6e-0644-43f8-a2b6-7fac399eadff" width="50%"/>

**Comment by DolomitesTom** [comment]
The nearby POIs are not shown during navigation, only the favorites. I have configured both to be shown on the left side. Is there a wrong setting?

**Reply by devemux86** [reply]
Nearby POI (the enabled ones) are displayed after starting navigation.

They are searched either offline (if offline POI are activated) or online.

**Reply by devemux86** [reply]
There is also a quick switch in navigation options (long press direction arrow at top-left).
Maybe you have it disabled?

**Reply by DolomitesTom** [reply]
Not in my display. They are only shown as blue dots or icons on the map but not in panels.

**Reply by DolomitesTom** [reply]
> There is also a quick switch in navigation options (long press direction arrow at top-left). Maybe you have it disabled?

Yes it was disabled. Is this also working in Simulation or only in navigation?

**Reply by devemux86** [reply]
> Is this also working in Simulation or only in navigation?

In both.

**Reply by DolomitesTom** [reply]
Then I still don't see the panels

**Reply by devemux86** [reply]
It works here, just tested it again.

It must be something in your settings or the activated POI or internet (if you don't use poi).
(navigation must start running and after a while the panels appear)

**Reply by DolomitesTom** [reply]
Yes seems to work after a while. At the moment I only can simulate, but I have the feeling, that sometimes, not always, it shows the same location twice (in two panels) and even if the virtual GPS is already far away the panels stay with 10m distance and don't get deleted.

**Reply by DolomitesTom** [reply]
![Screenshot_20240820_171707_Cruiser](https://github.com/user-attachments/assets/bd46b411-14b9-46f3-bb3a-56aceb716c76)
![Screenshot_20240820_172457_Cruiser](https://github.com/user-attachments/assets/3b97ec44-26fd-446d-a095-5f5c0ceda647)

Not only twice but even three times if three panels are configured for POIs

**Reply by devemux86** [reply]
Search can take a while if activate many large maps and for very long routes.

About the panels, e.g. fuel stations may exist in the 2 sides of the route at same location.
Better post examples to understand what you mean.

You can also select the side of route in search / advanced settings.

**Reply by devemux86** [reply]
Please post the locations and the route,
otherwise it is impossible to reproduce any report.

**Reply by DolomitesTom** [reply]
Can take a while... Wasn't long press on such a panel the function to insert this into the route? This doesn't work any more.

**Reply by devemux86** [reply]
You can also zoom in on POI locations on the map and tap on them to see how many there are.

**Reply by devemux86** [reply]
> Wasn't long press on such a panel the function to insert this into the route? This doesn't work any more.

It still works, obviously in navigation.

(simulation does not have rerouting)

**Reply by devemux86** [reply]
It depends on how many large maps (and poi) are activated, how long is the route.

Without more information we cannot know what happens, as it works here.

**Reply by devemux86** [reply]
Test first with a single map, a small route and a POI category that has POI near the route.

Also if you have changed the POI or search settings, create another app profile,
reset it to default settings and try with that.

**Reply by devemux86** [reply]
For example, if a route crosses France and Germany...

These countries have millions of POI, any search takes time.

**Reply by DolomitesTom** [reply]
Yes it seems that it takes a while with activated France and Germany POI file, but it finally works.

**Reply by devemux86** [reply]
@DolomitesTom thanks for the feedback. I will check how we can improve this.

Can you tell me:
- Whlch offline maps you activated?
- Which nearby POI you enabled?
- How long is the route you test?

**Reply by DolomitesTom** [reply]
Unfortunately I can't tell this anymore, since I just had an error when opening, all offline maps got deactivated and it switched to online maps (maybe to quick opened the app after reboot of the device?). But normally I have Austria, Belgium, France, Germany, Italy, Luxemburg and Switzerland active and POI France, Germany and Luxemburg.
Nearby POIs: fuel stations, Eat and Drink, Café, Biergarten, Supermarket
Route was 915 km, my next 2 day-trip

Question: the option for the distance is "Always", slider from 0.1 to 5 km. What does "always" mean?

**Reply by devemux86** [reply]
How do you use different offline maps and poi?
Do you activate the poi separately in the "Offline POI" screen with the extended menu?

> the option for the distance is "Always", slider from 0.1 to 5 km. What does "always" mean?

This is the distance from which the nearby panels are displayed.
"Always" means they appear always on the screen.

**Reply by DolomitesTom** [reply]
> Do you activate the poi separately in the "Offline POI" screen with the extended menu?

yes



> "Always" means they appear always on the screen.

in panels or additionally on the map? and till which distance?

**Reply by devemux86** [reply]
It is the display distance of panels before we reach the nearby POI.
Then these panels disappear and the next panels can be displayed.

**Reply by devemux86** [reply]
I still think that we do not need such complicated settings.

We can turn on / off the nearby panels in navigation's quick options.
(long press direction icon at top-left)

**Comment by DolomitesTom** [comment]
just made b9 crash. was in simulation and had selected another route from my routes.

**Reply by MotoUKRider** [reply]
@DolomitesTom Interesting. On my S20, I plotted a short route (through London), set it to "simulate" and as that was "navigating", I then loaded a local'ish GPX route and Cruiser automatically reverted to simulating that route, no issues.

**Comment by devemux86** [comment]
We can make a change in the nearby POI search in navigation:
- Nearby POI are now searched first in the activated offline poi databases.

- We can change this and if you have disabled the POI in search settings,
the nearby POI can be searched directly online, which will be faster.

**Reply by DolomitesTom** [reply]
Sounds good

**Comment by devemux86** [comment]
I wrote a new algorithm for the nearby POI that is dozens of times faster.

Testing **Bayern** map and populous "**Food**" POI on a long route of **484 km**:
- Old algorithm: **77 sec**

- New algorithm: **8 sec** :rocket:

**Reply by devemux86** [reply]
The good news is that this fast time remains the same even if we enable all nearby POI categories.

(the old algorithm got slower with more categories enabled)

The magic of mathematics! :magic_wand:

**Reply by Ciutnik** [reply]
> The magic of mathematics! 

Words of appreciation for the magician! 👍

**Comment by devemux86** [comment, marked as answer]
- You can select the nearby POI in navigation between all available POI categories in https://github.com/devemux86/cruiser/discussions/641
(in navigation settings)
